commit c4e2965a4b83f39927e1cfc47ff6941cac76afb7
Author: Sage Weil <sweil@redhat.com>
Date:   Thu Feb 14 18:21:38 2019 -0600

    mgr: hold GIL while generating a typed option value
    
    Drop the GIL while looking up the module and option value.  Retake
    it before calling get_typed_option_value(), where we generate the
    Python objects for the value.
    
    Fixes: http://tracker.ceph.com/issues/38123
    Fixes: http://tracker.ceph.com/issues/38204
    Fixes: http://tracker.ceph.com/issues/38292
    Signed-off-by: Sage Weil <sage@redhat.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index dca77dc6039..e90d877b852 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -529,19 +529,20 @@ PyObject *ActivePyModules::get_typed_config(
   const std::string &module_name,
   const std::string &key) const
 {
+  PyThreadState *tstate = PyEval_SaveThread();
   std::string value;
   bool found = get_config(module_name, key, &value);
   if (found) {
     PyModuleRef module = py_module_registry.get_module(module_name);
+    PyEval_RestoreThread(tstate);
     if (!module) {
         derr << "Module '" << module_name << "' is not available" << dendl;
         Py_RETURN_NONE;
     }
-
     dout(10) << __func__ << " " << key << " found: " << value << dendl;
     return module->get_typed_option_value(key, value);
   }
-
+  PyEval_RestoreThread(tstate);
   dout(4) << __func__ << " " << key << " not found " << dendl;
   Py_RETURN_NONE;
 }
